### ISSUE TITLE
How to change the trigger type

### ISSUE THREAD

**Issue by Lemon2911** [naqslab/naqslab_devices#2]
Hello,I want to add the novatech 409C to the list of devices. 
however, unlike the 409B, the 409C uses a rising edge to trigger the table mode.I don't know how to modify the code to do this, I hope to get your help

**Comment by dihm** [comment]
I had no idea that novatech finally released a new model, and with much better functionality! We should definitely add support.

For triggering, I don't know that anything actually needs to change. It's been a while since I've looked at all this, but I recall that the current trigger for the 409B-AC presumes rising edge. In synchronous mode, this is actually correct (but requires modification of the device to bridge pins 10 and 14 via a buffer). In asynchronous mode, you get reasonable behavior if you shorten the trigger pulses to their minimum width.

More annoyingly (but in a good way), you'll need to modify things to allow all four channels to have table mode, and to alter the manual external clock frequency handling. It should actually make the code quite a bit simpler since there is a lot of special case handling regarding both of those things.

I'm certainly interested in getting this kind of support going. I recently developed basically a [home-made version](https://github.com/qtc-umd/dds-sweeper) of the 409C which has the same basic specs and functionality (plus native linear sweeps), but never got around to writing the labscript driver. Supporting either device will make supporting the other fairly trivial.

**Comment by Lemon2911** [comment]
Yes, the 409C can handle external clock scaling automatically and has a shorter table dwell time. In addition, it supports higher baud rates (up to 460800), which I believe can better solve the problem of long wait times when writing large number of table commands.

It was imperative that I currently figure out how LabScript Suite handles triggering, i.e., different devices may require different trigger conditions, some triggering on the rising edge, while others on the falling edge.

I read the code of the 409B, and it only mentions that the device needs to be sent different commands to make the device work in different trigger modes, but what puzzles me is how the labscript suite generates the correct time series based on the different trigger modes?

**Comment by dihm** [comment]
How labscript handles the triggering of the 409B (or equivalently the DDS9m) is not well documented, but there are a few threads on the list serve that go into much more detail. [This one](https://groups.google.com/g/labscriptsuite/c/Bf4UJMgmky0/m/xI0-q5x7AAAJ) in particular is probably the most descriptive. In any case, I'm quite confident that, as far as the driver is concerned, the 409B uses rising edge triggering. Making that work in practice has always required solutions outside of the driver itself, which are described in that thread.

So in short, I think you are basically good to go, as is.

**Comment by Lemon2911** [comment]
Ok, I'm working on trying to adapt to 409C right now.Thanks for the reply
